python/shotgun_data/shotgun_data_retriever.py

```diff
@@ -596,29 +680,51 @@ def _download_url(self, file_path, url, entity_type, entity_id, field):
                           the url needs to be requeried, this field will be
                           where the fresh url is pulled from.
+        :returns: Full path the downloaded file. This value may be different
+                  than the input `file_path` if the resolved url's extension
+                  differed from what was specified.
         """
         try:
+            download_path = sgtk.util.download_url(
+                self._bundle.shotgun, url, file_path, True
+            )
+            file_path = download_path
+        except TypeError:
+            # Probably an older version of core that doesn't have the
+            # sgtk.util.download_url() `use_url_exension` arg implemented.
+            # Try running without that arg value
             sgtk.util.download_url(self._bundle.shotgun, url, file_path)
+
         except TankError, e:
-            sg_data = self._bundle.shotgun.find_one(
-                entity_type,
-                [["id", "is", entity_id]],
-                [field],
-            )
+            if field is not None:
+                sg_data = self._bundle.shotgun.find_one(
+                    entity_type,
+                    [["id", "is", entity_id]],
+                    [field],
+                )
 
-            if sg_data is None or sg_data.get(field) is None:
-                # This means there's nothing in Shotgun for this field, which
-                # means we can't download anything.
-                raise IOError(
-                    "Field %s does not contain data for %s (id=%s)." % (
-                        field,
-                        entity_type,
-                        entity_id,
+                if sg_data is None or sg_data.get(field) is None:
+                    # This means there's nothing in Shotgun for this field, which
+                    # means we can't download anything.
+                    raise IOError(
+                        "Field %s does not contain data for %s (id=%s)." % (
+                            field,
+                            entity_type,
+                            entity_id,
+                        )
                     )
-                )
-            else:
-                url = sg_data[field]
-                sgtk.util.download_url(self._bundle.shotgun, url, file_path)
+                else:
+                    url = sg_data[field]
+                    try:
+                        download_path = sgtk.util.download_url(
+                            self._bundle.shotgun, url, file_path, True
+                        )
+                        file_path = download_path
+                    except TypeError:
+                        # Probably an older version of core that doesn't have the
+                        # sgtk.util.download_url() `use_url_exension` arg implemented.
+                        # Try running without that arg value.
+                        sgtk.util.download_url(self._bundle.shotgun, url, file_path)
```

Comment:
for the legacy fallbacks, do we want to append a hard coded jpg file extension? Or is that bad? Wasn't sure.